python/ray/rllib/agents/impala/vtrace.py

```diff
@@ -32,204 +33,209 @@
+
+    with tf.name_scope(name, values=[behaviour_policy, target_policy, actions,
+                                     discounts, rewards, values,
+                                     bootstrap_value]):
+        target_action_policy = select_policy_values_using_actions(
+            target_policy, actions)
+        behaviour_action_policy = select_policy_values_using_actions(
+            behaviour_policy, actions)
+
+        rhos = get_rhos(target_action_policy, behaviour_action_policy)
+
+        vtrace_returns = _from_importance_weights(
+            rhos=rhos,
             discounts=discounts,
             rewards=rewards,
             values=values,
             bootstrap_value=bootstrap_value,
             clip_rho_threshold=clip_rho_threshold,
             clip_pg_rho_threshold=clip_pg_rho_threshold)
+
         return VTraceFromLogitsReturns(
-            log_rhos=log_rhos,
-            behaviour_action_log_probs=behaviour_action_log_probs,
-            target_action_log_probs=target_action_log_probs,
+            rhos=rhos,
+            behaviour_action_policy=behaviour_action_policy,
+            target_action_policy=target_action_policy,
             **vtrace_returns._asdict())
 
 
-def from_importance_weights(log_rhos,
-                            discounts,
-                            rewards,
-                            values,
-                            bootstrap_value,
-                            clip_rho_threshold=1.0,
-                            clip_pg_rho_threshold=1.0,
-                            name='vtrace_from_importance_weights'):
+def select_policy_values_using_actions(policy_logits, actions):
+    """
+    Computes action log-probs from policy logits and actions.
+
+    In the notation used throughout documentation and comments, T refers to the
+    time dimension ranging from 0 to T-1. B refers to the batch size and
+    NUM_ACTIONS refers to the number of actions.
+
+    Args:
+      policy_logits: A float32 tensor of shape [T, B, NUM_ACTIONS] with
+        un-normalized log-probabilities parameterizing a softmax policy.
+      actions: An int32 tensor of shape [T, B] with actions.
+
+    Returns:
+      A float32 tensor of shape [T, B] corresponding to the sampling log
+      probability of the chosen action w.r.t. the policy.
+    """
+
+    log_probs = []
+    for i in range(len(policy_logits)):
+        log_probs.append(-tf.nn.sparse_softmax_cross_entropy_with_logits(
+            logits=policy_logits[i], labels=actions[i]))
+
+    return log_probs
+
+
+def get_rhos(behaviour_action_log_probs, target_action_log_probs):
+    """With the selected policy values (logits or probs) subclasses compute
+    the rhos for calculating the vtrace."""
+    log_rhos = [t - b for t,
+                b in zip(target_action_log_probs, behaviour_action_log_probs)]
+    log_rhos = [tf.convert_to_tensor(l, dtype=tf.float32) for l in log_rhos]
+    log_rhos = tf.reduce_sum(tf.stack(log_rhos), axis=0)
```

Comment:
IIUC, the change to support MultiDiscrete is mainly combining the independent log probabilities. In this case, is it possible to move the summations outside of vtrace.py to minimize the changes?

Reply:
Correct. The summation of the independent log probabilities needs to be done inside vtrace.py since it comes after selecting policy values using actions for the behaviour and target policies and before its use in _from_importance_weights method (lines 119-133).